spec/index.md

```diff
@@ -3702,6 +3732,72 @@ The functions `query_as_update` and `system_task_as_update` turns a query functi
 
 Note that by construction, a query function will either trap or return with a response; it will never send calls, and it will never change the state of the canister.
 
+#### Spontaneous request rejection {#request-rejection}
+
+The system can reject a request at any point in time, e.g., because it is overloaded. The following specification is an overapproximation of the rejection behavior. In particular, the system is guaranteed not to emit `DESTINATION_INVALID` in some cases, e.g., if the callee exists, and if the caller successfully calls the caller once, subsequent calls will not fail with `DESTINATION_INVALID`.
+
+Condition:
+```html
+S.messages = Older_messages · CallMessage CM · Younger_messages
+(CM.queue = Unordered) or (∀ msg ∈ Older_messages. msg.queue ≠ M.queue)
+reject_code ∈ { SYS_FATAL, SYS_TRANSIENT, DESTINATION_INVALID }
+```
+
+State after (given some `reject_msg`):
+```html
+S.messages =
+    Older_messages ·
+    ResponseMessage {
+        origin = CM.origin;
+        response = Reject (reject_code, reject_msg);
+        refunded_cycles = CM.transferred_cycles;
```

Comment:
Can the system really refund the cycles here in all cases, e.g., couldn't there be a spontaneous reject corresponding to a request that has already made it into a state where it is unknown whether it has been processed or not?

Reply:
I guess it's fine to refund cycles (everywhere) as long as we only accept at most a single refund for every downstream call.

Reply:
Not sure I follow. If the receiver gets to see the request then it could technically accept the cycles. If we also refund them then we have created new cycles, no?

Reply:
For the specific error codes above (SYS_FATAL, SYS_TRANSIENT, DESTINATION_INVALID), the receiver must not have received the request.

Reply:
I see makes sense - overlooked the restriction to certain error cases. 

Do we want to specify things this way? It might be a simplification worth considering to initially never refund cycles when we see errors at all and only start refunding in a later iteration.

Reply:
> I bundled the fix to this with the best-effort PR since I was too lazy to handle merge conflicts :)

Could you please create a separate PR after all?

Reply:
Now that the deadline has been moved to origin, there won't be a large merge conflict if we do this in a separate PR, so I can do that.